Add `AnimatableTableViewController` with designable `UIRefreshControl`

phimage opened the pull request:
and other features same as `AnimatableViewController`

`RefreshControlDesignable` allow to custom the internal `UIRefreshControl`

`UIRefreshControl` must not be set in subview but using refreshControl attribute of `UITableViewController`


ReStart of PR #418 due to many conflict
Modification since review by @tbaranes and @JakeLin 
- remove some space and add missing open on IBoutlet variable copied from AnimatableViewController
- add `TimeLineTableViewController` in demo app , to show how refresh control could be customised (simulate the refresh using `DispatchQueue`)
- fix a bug : if view not already loaded. If view not loaded, refreshControl could not be set

e-marchand committed:
Add `AnimatableTableViewController` with designable `UIRefreshControl`

tbaranes commented on CHANGELOG.md:
```diff
@@ -16,13 +16,15 @@ All notable changes to this project will be documented in this file.
-
+- Add `AnimatableTableViewController` to support `UIRefreshControl` customization.
+[#418](https://github.com/IBAnimatable/IBAnimatable/pull/418) by [@phimage](https://github.com/phimage)
+CHANGELOG
```
There's a `CHANGELOG` to remove here!

tbaranes commented on Documentation/APIs.md:
```diff
@@ -170,6 +171,26 @@ Easily add color layer on top of the UI element especially `AnimatableImageView`
 | hideNavigationBar | Bool | whether to hide navigation bar. The default value is `false`. |
 
+#### `RefreshControlDesignable`
+Display a `UIRefreshControl` in `AnimatableTableViewController`
```
I wouldn't add this line, I think the protocol name is clear enough (moreover, we may conform to this protocol in other classes).

tbaranes commented on IBAnimatable/AnimatableTableViewController.swift:
```diff
@@ -0,0 +1,111 @@
+//
+//  AnimatableTableViewController.swift
```
You can remove these two lines. Take a look at the other headers of the project, we are just keeping the author lines.

tbaranes commented on IBAnimatable/RefreshControlerDesignable.swift:
```diff
@@ -0,0 +1,41 @@
+public extension RefreshControlDesignable where Self: UITableViewController {
+
+  func configureRefreshController() {
+    if isViewLoaded {
```
Guard?

phimage replied:
agree

tbaranes commented on IBAnimatableApp/TimeLineTableViewController.swift:
```diff
@@ -0,0 +1,63 @@
+    // Simulate an asynchrone refresh, could be a network request...
+    let time: TimeInterval = 5
+    self.updateMessage(refreshControl: refreshControl, time: time)
+    DispatchQueue.background.after(time) {
```
Why not directly running this on `main`?

phimage replied:
To simulate a good practice, doing a  job in background like get data from network

tbaranes commented on IBAnimatableApp/TimeLineTableViewController.swift:
```diff
@@ -0,0 +1,63 @@
+
+  // Function to update message recursively
+  func updateMessage(refreshControl: UIRefreshControl, time: TimeInterval) {
+    if time < 0 {
```
guard?

phimage replied:
ummm here not agree `if` is shorter and more understandable  but I will do it

tbaranes reviewed (changes requested):
I let you a few more minor comments, if you can take a look at them... :)

Also, a new swiftlint warning appeared since we enabled `line_length`, can you take care of it? If you want to see the warnings directly in Xcode, you just have to install [SwiftLint](https://github.com/realm/SwiftLint)

tbaranes commented:
One more thing that we should have precise it earlier, my apologies. We don't add features to the `Demo.storyboard`, but directly in the `Playground` part which is a showoff of all the features available in `IBAnimatable`.

For the refresh control, you could add an entry either in:

- `Playground` -> `User Interface` -> `Refresh Control`
- `Playground` -> `RefreshControl`  (new) -> `UITableView` 
- Third place?

Both cases are ok, it's up to you, maybe depending if we want to add an example for the `collectionView` (when we will support it)

Sorry for not letting you know about this before 😟

phimage commented:
I use swiftlint and update it
I have missing one

e-marchand committed:
Remove from demo app the `AnimatableViewController`
Use guard instead of if when checking if view is loaded before configuring the table view refresh control
Remote useless documentation line
Fix change log

e-marchand committed:
Fix swiftlint warning on line length for AnimatableTableViewController

phimage commented:
I have no time to make new storyboards in playground 
And I have no internet at home currently, could merge without the demo?

tbaranes reviewed (approved):
Let's wait for @JakeLin review, but  I'm ok to merge it without the demo. We can add it later! Thanks for taking care of the comments.

JakeLin commented on IBAnimatable/AnimatableTableViewController.swift:
```diff
@@ -0,0 +1,109 @@
+  @IBInspectable open var rootWindowBackgroundColor: UIColor?
+
+  // MARK: - TransitionAnimatable
+  @IBInspectable open var _transitionAnimationType: String? {
```
I think we don't want that to be `open` and let it only visible in Interface Builder. We can remove `open` and let it as  the default `internal`.

JakeLin commented on IBAnimatable/AnimatableTableViewController.swift:
```diff
@@ -0,0 +1,109 @@
+  @IBInspectable open var transitionDuration: Double = .nan
+
+  open var interactiveGestureType: InteractiveGestureType = .none
+  @IBInspectable open var _interactiveGestureType: String? {
```
Same comment as above.

JakeLin commented on IBAnimatableApp/RefreshTableViewController.swift:
```diff
@@ -0,0 +1,60 @@
+import UIKit
+import IBAnimatable
+
+public class RefreshTableViewController: AnimatableTableViewController {
```
Nice demo app 👍

JakeLin reviewed (approved):
👍 Well done @phimage

JakeLin committed:
Make `_transitionAnimationType` and `_interactiveGestureType` `internal` and they can be only used in Interface Builder.

JakeLin commented:
@phimage I have fixed my comment and merged the PR now 🎉🎉🎉. Thank you very much for creating such a great feature for IBAnimatable. It makes our life easier to configure the refersh control in Interface Builder.

For `@IBInspectable var _transitionAnimationType: String?`, because Interface Builder can't support enum. And we want to have nice programmatical API with strongly type like `open var transitionAnimationType: TransitionAnimationType = .none`. To do that, we found out a workaround, we use `String` as the type to make the inspectable property can be configured in Interface Builder. But don't want the user/client code to see `_transitionAnimationType`, then we make them as `internal`. In that case, we can configure `transitionAnimationType` in Interface Builder and use it in code as an enum. More info can be found in #218 @lastMove and @tbaranes did a great job on that.

tbaranes commented:
For your information, I added an entry in #238 in order to not forget that we have to update the playground storyboard in order to have a demo for refresh control.
